### ISSUE TITLE
How to use cosine proximity?

### ISSUE THREAD

**Issue by maxbry** [fchollet/keras#3031]
I roughly have this network structure:

```
input_a = Input(shape=(input_dim, 1))
input_b = Input(shape=(input_dim, 1))

vec_a = embedding_a(input_a)
vec_b = embedding_b(input_b)

cos_distance = merge([vec_a, vec_b], mode='cos', dot_axes=1)
cos_distance = Reshape((1,))(cos_distance)

model = Model(input=[input_a, input_b], output=[cos_similarity])
model.compile(optimizer='sgd', loss='cosine_proximity')
```

When training this network, loss gets nan. Do I use cosine_proximity correctly? Also, cosine proximity/distance is ranged between -1 and 1. Do I have to use this as targets or 1 and 0 as usually?

By the way: The name does not seem intuitive because we usually want to minimize what we define as loss. Shouldn't it be cosine_distance?

**Comment by dr-costas** [comment]
Any news on that? 

I have exactly the same questions and remark about the name of the objective. Regarding the latter, I think that the return should be `1 - K.mean(y_true * y_pred, axis=-1) ` instead of `-K.mean(y_true * y_pred, axis=-1)` and the name should be _cosine distance_ instead of _proximity_ since the objective is to minimize the distance and not the proximity.

**Comment by fchollet** [comment]
> 1 - K.mean(y_true * y_pred, axis=-1) instead of -K.mean(y_true * y_pred, axis=-1)

There will be no difference in behavior between the two, what you propose is just the offset the same quantity by one.

**Comment by dr-costas** [comment]
For those that have the same issue:

```
from keras import backend as K


def cos_distance(y_true, y_pred):
    def l2_normalize(x, axis):
        norm = K.sqrt(K.sum(K.square(x), axis=axis, keepdims=True))
        return K.maximum(x, K.epsilon()) / K.maximum(norm, K.epsilon())
    y_true = l2_normalize(y_true, axis=-1)
    y_pred = l2_normalize(y_pred, axis=-1)
    return 1-K.mean(y_true * y_pred, axis=-1)
```

Thnx @Js-Mim

**Comment by anMabbe** [comment]
I would not advise to use this function as the results will significatively differ from what is expected, try it on some randomly generated values to see the difference : 

```python

def cos_distance(y_true, y_pred):
    def l2_normalize(x, axis):
        norm = K.sqrt(K.sum(K.square(x), axis=axis, keepdims=True))
        return K.maximum(x, K.epsilon()) / K.maximum(norm, K.epsilon())
    y_true = l2_normalize(y_true, axis=-1)
    y_pred = l2_normalize(y_pred, axis=-1)
    return K.mean(y_true * y_pred, axis=-1)


  
for i in range(5):
    x=np.random.randn(2)
    y=np.random.randn(2)
    a=cosine_proximity(x,y)
    b=-cos_distance(x,y)
    print a.eval()
    print b.eval()
    print "\n"

```

**Comment by anMabbe** [comment]
By the way, does someone know why we should modify the numerator this way ? I thought that only the numerator would give rise to nan-related errors.

**Comment by alfiya400** [comment]
@anMabbe  Hi, I'm new to keras, but if cosine_proximity and cosine_distance are the same then we also need to add `K.sum` to make an average cosine distance across all data points:
```
def cos_distance(y_true, y_pred):
    y_true = K.l2_normalize(y_true, axis=-1)
    y_pred = K.l2_normalize(y_pred, axis=-1)
    return K.mean(1 - K.sum((y_true * y_pred), axis=-1))
```
I interpret it as an average cosine distance between actual output and prediction.

Check:
```
# cos_distance = [1, 1, 0] => avg = 2/3
x1 = np.array([[0, 1], [1, 0], [1, 1]])
x2 = np.array([[1, 0], [0, 1], [1, 1]])
print(K.eval(cos_distance(x1, x2)))
```
Output:
```
0.6666666666666666
```

**Comment by kevindewalt** [comment]
@alifya400 I had to use 
```
x1 = K.variable(np.array([[0, 1], [1, 0], [1, 1]]))
x2 = K.variable(np.array([[1, 0], [0, 1], [1, 1]]))
```
Keras 1.2.2
